### ISSUE TITLE
[Proposal] Add SQL id, request logs, and metrics

### ISSUE THREAD

**Issue by gaodayue** [apache/incubator-druid#6301]
# Motivation

Druid SQL should be able to emit SQL metrics and write SQL request logs. It should also help users correlate SQL request with underlying native queries so that users can troubleshoot SQL performance issue using native query metrics.

This is a follow-up work of [#4047](https://github.com/apache/incubator-druid/issues/4047). I'm trying to implement all the missing functionalities except SQL cancellation.

# Proposed implementation

There are two entries for SQL processing: `SqlResource` for HTTP request and `DruidStatement` for JDBC request. In order to remove duplicated logic, `SqlLifecycle` is added as a central place to manage sql execution. It follows the same design as `QueryLifecycle` except it's thread-safe, since `SqlLifecycle` can be invoked in multi-threads in JDBC use cases.

The public interface of SqlLifecycle would be

```java
class SqlLifecyle {
  // state transition: NEW -> INITIALIZED
  void initialize(SqlQuery);
  void initialize(String sql, Map<String, Object> queryContext);
  // state transition: INITIALIZED -> PLANNED
  void plan(AuthenticationResult);
  void plan(HttpServletRequest);
  // state transition: PLANNED -> AUTHORIZED or UNAUTHORIZED
  Access authorize();
  // state transition: AUTHORIZED -> EXECUTING
  Sequence<Object[]> execute();
  // state transition: -> DONE
  void emitLogsAndMetrics(Throwable, String remoteAddress, long bytesWritten);
}
```

In `initialize` method, a `sqlId` is assigned to each SQL query. User can specify `sqlId` in query context, or Druid will generate a UUID for it, just like `queryId` for native query. We call it `sqlId` to make it clear that `sqlId` is for SQL request while `queryId` is for native query, and one SQL can map to several native queries.

To help users correlate SQL with native queries, `sqlId` is added to query context of native queries, so that from native query's request log, it's clear whether the native query is from end user or SQL, and if it's from SQL, which SQL is it comes from. Similarly, `queryId` of all native queries underpinning a SQL are collected and stored in SQL's query context key `nativeQueryIds`, so that from SQL's request log or metrics, it's clear which native queries it contains. Finally, `sqlId` and its corresponding `queryId`s are returned in `X-Druid-SQL-Id` and `X-Druid-Native-Query-Ids` HTTP headers.

Two SQL metrics are emitted in `emitLogsAndMetrics` method, namely `sql/time` and `sql/bytes`. I perfer `sql/time` to `query/sql/time` proposed in [#4047](https://github.com/apache/incubator-druid/issues/4047) because I think in the past, all metrics starting with "query/" are for native queries, and we should stick to that convention. Also `sql/time` makes SQL a first citizen and is more intuitive for me to understand. SQL metrics will have the following dimensions: id, nativeQueryIds, dataSource, remoteAddress, and success.

SQL requests are logged in `emitLogsAndMetrics` via `SqlRequestLogger`. Several built-in impls such as `NoopSqlRequestLogger` and `FileSqlRequestLogger` are provided, of cause we can add more later. `SqlRequestLogger` is configured via `druid.sql.request.logging`, and can be extended in similar way as RequestLogger.
```java
public interface SqlRequestLogger
{
  void log(SqlRequestLogLine sqlRequestLogLine) throws IOException;
  default void start() throws IOException {}
  default void stop() {}
}
```

**Comment by gianm** [comment]
Hi @gaodayue, sorry it took a while to get you feedback on this proposal. I think it looks great. I have some small suggestions:

- `sqlQueryId` and `sqlQuery/time` look better to me than `sqlId` and `sql/time`. It emphasizes that they are, in fact, queries (just SQL instead of native). I do agree with you that either one of these makes more sense than `query/sql/time`.
- I think it's better to add a method to RequestLogger that accepts a SQL query, instead of creating a new SqlRequestLogger. I think, on balance, it will be nicer for people to get all their logs in one place rather than having to worry about configuring two different things.
- Maybe this is a good opportunity to fix #4826 as well. My thoughts on that are in https://github.com/apache/incubator-druid/issues/4826#issuecomment-330560830.
- The `sqlQueryId` sounds like a natural choice for implementing a SQL query cancelation API in the future - do you agree?

**Comment by gaodayue** [comment]
Hi @gianm , thanks for your suggestions.

> sqlQueryId and sqlQuery/time look better to me than sqlId and sql/time. It emphasizes that they are, in fact, queries (just SQL instead of native). I do agree with you that either one of these makes more sense than query/sql/time.

Sure, sqlQueryId and sqlQuery/time express more accurate meaning. I'll also change `X-Druid-SQL-Id` to `X-Druid-SQL-Query-Id`.

> I think it's better to add a method to RequestLogger that accepts a SQL query, instead of creating a new SqlRequestLogger. I think, on balance, it will be nicer for people to get all their logs in one place rather than having to worry about configuring two different things.

I think adding a separate `SqlRequestLogger` is more flexible than reusing `RequestLogger`, for example

- It allows users to enable sql logging while disabling native query logging, which can be useful in cases where only SQL is used.
- It allows users to choose different implementations for SQL and native query logging. For instance, users may want to log sql request to both kafka and files, while only keep native query logs in local files for diagnostic purpose.

> Maybe this is a good opportunity to fix #4826 as well. My thoughts on that are in #4826 (comment).

Yeah, but I think it's better to address that issue in a separate PR for backport purpose. Let me finish this first :)

> The sqlQueryId sounds like a natural choice for implementing a SQL query cancelation API in the future - do you agree?

Totally agree.
